crates/ruma-federation-api/src/authenticated_media.rs

```diff
@@ -0,0 +1,291 @@
+
+    let metadata_raw = serde_json::to_vec(&metadata)?;
+    let metadata_part = MimePart::new("application/json", metadata_raw)
+        // We need to specify the `Content-Transfer-Encoding` otherwise it is base64-encoded.
+        .transfer_encoding("binary");
+
+    let content_part = match content {
+        FileOrLocation::File(content) => MimePart::new(content.content_type, content.file)
+            .header("Content-Disposition", Text::new(content.content_disposition)),
+        FileOrLocation::Location(location) => {
+            MimePart::raw(Vec::<u8>::new()).header("Location", Text::new(location))
+        }
+    }
+    // We need to specify the `Content-Transfer-Encoding` otherwise it is base64-encoded.
+    .transfer_encoding("binary");
+
+    let body_part = MimePart::raw(vec![metadata_part, content_part]);
+    let mut body_full = Vec::new();
+    body_part.write_part(&mut body_full).expect("writing to a Vec should never fail");
+
+    // The Content-Type should be the only header included at the beginning of the body.
+    let mut header_end_pos = memchr::memmem::find(&body_full, b"\r\n")
+        .expect("there should always be a least one newline");
+    // Remove `Content-Type: ` at the start of the header.
+    let mut content_type = body_full[14..header_end_pos].to_vec();
+
+    // If the next line starts with a whitespace, it's still the same header.
+    header_end_pos += 2;
+    while body_full.get(header_end_pos).is_some_and(|c| c.is_ascii_whitespace()) {
+        // Skip the whitespaces at the beginning of the line, but stop if we reach a newline.
+        let (start_line_pos, _) = body_full[header_end_pos..]
+            .iter()
+            .enumerate()
+            .find(|(_pos, c)| **c == b'\r' || !c.is_ascii_whitespace())
+            .expect("a header should end before the end of the body");
+        header_end_pos += start_line_pos;
+
+        let end_line_pos = memchr::memmem::find(&body_full[header_end_pos..], b"\r\n")
+            .expect("a header should end with a newline");
+        let line = &body_full[header_end_pos..header_end_pos + end_line_pos];
+
+        // Append the new header line.
+        if !line.is_empty() {
+            content_type.extend_from_slice(line);
+        }
+
+        header_end_pos += end_line_pos + 2;
+    }
+
+    let mut body = T::default();
+    body.put(&body_full[header_end_pos..]);
+
+    Ok(http::Response::builder().header(http::header::CONTENT_TYPE, content_type).body(body)?)
+}
+
+/// Deserialize the given metadata and content from a `http::Response` with a `multipart/mixed`
+/// body.
+#[cfg(feature = "server")]
+fn try_from_multipart_mixed_response<T: AsRef<[u8]>>(
+    http_response: http::Response<T>,
+) -> Result<
+    (ContentMetadata, FileOrLocation),
+    ruma_common::api::error::FromHttpResponseError<ruma_common::api::error::MatrixError>,
+> {
+    use mail_parser::{parsers::MessageStream, GetHeader, HeaderName, HeaderValue, MessageParser};
```

Comment:
This library feels very half-baked from a correctness perspective, for example `skip_crlf()` actually skips any number of whitespace characters until it finds a `\n`, which sounds ripe for header parsing confusion. From a quick look around I found https://crates.io/crates/mime_multipart, have you evaluated it? At least the types look like they provide what we need.

Reply:
What you call "half-baked", I would call permissive. This is a library used in production so it probably is more lenient with the data it receives to be able to accommodate small deviations from the RFC. However I don't see how this particular implementation would cause a problem if it's only called in the right places.

I ended up using these libraries because, in my opinion, none of the alternatives are better, and mail-parser at least claims to have been tested against real emails and to be used in production. The main problem with other libraries, like mime_multipart, is that they usually depends on a full HTTP implementation, when we need them to only depend on the http crate.

mime_multipart is even more problematic that it depends on a very old version of hyper (the last compatible version was released 7 years ago) and I don't like the fact that it uses files for attachments because we need to read from them to get the data and is more error-prone than writing directly to a buffer.

Given the number of crates available, it looks like it would have been simpler for us if the MSC chose to use `multipart/form-data`.

Reply:
This library is no-longer used.